crates/topos-sequencer-subnet-runtime-proxy/src/subnet_runtime_proxy.rs

```diff
@@ -41,29 +42,60 @@ impl SubnetRuntimeProxy {
-            mpsc::unbounded_channel::<SubnetRuntimeProxyCommand>();
+        let (command_sender, mut command_rcv) = mpsc::channel::<SubnetRuntimeProxyCommand>(256);
         let ws_runtime_endpoint = format!("ws://{}/ws", &config.endpoint);
         let http_runtime_endpoint = format!("http://{}", &config.endpoint);
         let subnet_contract_address = Arc::new(config.subnet_contract_address.clone());
         let (command_task_shutdown_channel, mut command_task_shutdown) =
             mpsc::channel::<oneshot::Sender<()>>(1);
         let (block_task_shutdown_channel, mut block_task_shutdown) =
             mpsc::channel::<oneshot::Sender<()>>(1);
+        let (source_head_certificate_id_sender, source_head_certificate_id_received) =
+            oneshot::channel();
+
+        let certification = certification::Certification::new(
+            &config.subnet_id,
+            None,
+            config.verifier,
+            signing_key.clone(),
+        )?;
 
         let runtime_proxy = Arc::new(Mutex::from(Self {
             commands_channel: command_sender,
             events_subscribers: Vec::new(),
             config: config.clone(),
             command_task_shutdown: command_task_shutdown_channel,
             block_task_shutdown: block_task_shutdown_channel,
+            certification: certification.clone(),
+            source_head_certificate_id_sender: Some(source_head_certificate_id_sender),
         }));
 
         // Runtime block task
         {
             let runtime_proxy = runtime_proxy.clone();
             let subnet_contract_address = subnet_contract_address.clone();
             tokio::spawn(async move {
+                // To start producing certificates, we need to know latest delivered or pending certificate from TCE
+                // It could be also genesis certificate (also retrievable from TCE)
+                // Wait until we acquire first certificate for this network
+                let last_certificate_id = certification.lock().await.last_certificate_id;
+                if last_certificate_id.is_none() {
+                    info!("Waiting for the source head certificate id to continue with certificate generation");
+                    match source_head_certificate_id_received.await {
+                        Ok(last_certificate_id) => {
+                            info!(
+                                "Source head certificate id received {:?}",
+                                last_certificate_id.map(|id| id.to_string())
+                            );
+                            let mut certification = certification.lock().await;
+                            certification.last_certificate_id = last_certificate_id;
+                        }
+                        Err(e) => {
+                            panic!("Failed to get source head certificate, unable to proceed with certificate generation: {e}")
+                        }
+                    }
+                }
```

Comment:
@Freyskeyd Well, I intentionally did not want to lock `certification` when picking its `last_certificate_id` as it is used in other tasks and there is no chance of race because `oneshot` channel is used to set `certification.last_certificate_id` . Source head certificate is set only once, either manually in configuration when creating `subnet_runtime_proxy` or if it is not set at that point then waiting for `source_head_certificate_id` return value in app setup of the TCE component.

Reply:
Reviewing now the logic, it make sense to forbid usage of certification anywhere else before `last_certificate_id` is set.

Reply:
Actually not - task to get new blocks from subnet should continue working even if `last_certificate_id` is not set, just generating certificates should be skipped until it is set.